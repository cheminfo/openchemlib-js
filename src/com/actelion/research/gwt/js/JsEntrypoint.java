package com.actelion.research.gwt.js;

import com.google.gwt.core.client.EntryPoint;

public class JsEntrypoint implements EntryPoint {
  @Override
  public void onModuleLoad() {}
}
