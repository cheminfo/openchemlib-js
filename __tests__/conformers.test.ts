import { assert, beforeAll, describe, expect, it } from 'vitest';

import { ConformerGenerator, Molecule, Resources } from '../lib/index.js';

beforeAll(() => {
  Resources.registerFromNodejs();
});

describe('ConformerGenerator', () => {
  it('should return one conformer', () => {
    const mol = Molecule.fromSmiles('COCCON');

    const gen = new ConformerGenerator(1);
    expect(gen.getConformerCount()).toBe(0);

    expect(mol.getAtomZ(0)).toBe(0);
    const result = gen.getOneConformerAsMolecule(mol);
    expect(gen.getConformerCount()).toBe(1);
    expect(result).toBe(mol);
    expect(mol.getAtomZ(0)).not.toBe(0);
  });

  it('should generate conformers', () => {
    const mol = Molecule.fromSmiles('COCCON');

    const gen = new ConformerGenerator(1);
    expect(gen.getPotentialConformerCount()).toBe(1);
    expect(gen.getConformerCount()).toBe(0);
    gen.initializeConformers(mol);
    expect(gen.getPotentialConformerCount()).toBeGreaterThan(1);

    const conf1 = gen.getNextConformerAsMolecule();
    expect(gen.getConformerCount()).toBe(1);
    assert(conf1);
    const conf1Molfile = conf1.toMolfile();
    expect(conf1Molfile).toMatchSnapshot();

    const conf2 = gen.getNextConformerAsMolecule();
    expect(gen.getConformerCount()).toBe(2);
    assert(conf2);
    const conf2Molfile = conf2.toMolfile();
    expect(conf2Molfile).toMatchSnapshot();

    expect(conf2).not.toBe(conf1);
    expect(conf2Molfile).not.toBe(conf1Molfile);
  });

  it('should use and return the passed molecule', () => {
    const mol = Molecule.fromSmiles('COCCON');

    const gen = new ConformerGenerator(1);
    gen.initializeConformers(mol);

    expect(mol.getAtomZ(0)).toBe(0);

    const conf = gen.getNextConformerAsMolecule(mol);
    expect(conf).toBe(mol);
    expect(mol.getAtomZ(0)).not.toBe(0);
  });

  it('should molecules() return an iterator', () => {
    const mol = Molecule.fromSmiles('COCCON');

    const gen = new ConformerGenerator(1);
    gen.initializeConformers(mol);

    const allConformers = [...gen.molecules()];
    expect(allConformers).toHaveLength(26);

    // All molecules should be different
    expect(new Set(allConformers).size).toBe(26);

    expect(allConformers[0].getAtomZ(0)).not.toBe(allConformers[1].getAtomZ(1));

    expect(allConformers.every((conf) => conf instanceof Molecule)).toBe(true);
  });
});
