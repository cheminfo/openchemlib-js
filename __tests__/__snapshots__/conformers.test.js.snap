// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConformerGenerator > should generate conformers 1`] = `
"#1
Actelion Java MolfileCreator 1.0

 15 14  0  0  0  0  0  0  0  0999 V2000
    1.6038    1.1775   -1.7436 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.6164   -0.1245   -1.4739 O   0  0  0  0  0  0  0  0  0  0  0  0
    1.7449   -1.4436   -0.9286 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.5103   -1.3101    0.8709 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.4916   -0.0258    1.2038 O   0  0  0  0  0  0  0  0  0  0  0  0
   -1.0764   -0.4433    0.8311 N   0  0  0  0  0  0  0  0  0  0  0  0
    0.8502    0.9124   -2.5818 H   0  0  0  0  0  0  0  0  0  0  0  0
    2.2308    2.0933   -2.0679 H   0  0  0  0  0  0  0  0  0  0  0  0
    1.0263    1.4325   -0.7736 H   0  0  0  0  0  0  0  0  0  0  0  0
    2.3222   -2.4184   -1.1672 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.7122   -1.4701   -1.4491 H   0  0  0  0  0  0  0  0  0  0  0  0
    2.5348   -1.1378    1.3795 H   0  0  0  0  0  0  0  0  0  0  0  0
    1.0432   -2.2915    1.2686 H   0  0  0  0  0  0  0  0  0  0  0  0
   -1.6645    0.3987    0.8315 H   0  0  0  0  0  0  0  0  0  0  0  0
   -1.4292   -1.1042    1.5348 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  3  4  1  0  0  0  0
  4  5  1  0  0  0  0
  5  6  1  0  0  0  0
  1  7  1  0  0  0  0
  1  8  1  0  0  0  0
  1  9  1  0  0  0  0
  3 10  1  0  0  0  0
  3 11  1  0  0  0  0
  4 12  1  0  0  0  0
  4 13  1  0  0  0  0
  6 14  1  0  0  0  0
  6 15  1  0  0  0  0
M  END
"
`;

exports[`ConformerGenerator > should generate conformers 2`] = `
"#2
Actelion Java MolfileCreator 1.0

 15 14  0  0  0  0  0  0  0  0999 V2000
    2.8495   -0.2172   -3.1263 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.6164   -0.1245   -1.4739 O   0  0  0  0  0  0  0  0  0  0  0  0
    1.7449   -1.4436   -0.9286 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.5103   -1.3101    0.8709 C   0  0  0  0  0  0  0  0  0  0  0  0
    2.9853   -1.4939    1.6379 O   0  0  0  0  0  0  0  0  0  0  0  0
    2.7789   -1.3729    3.2856 N   0  0  0  0  0  0  0  0  0  0  0  0
    3.4828   -1.1545   -3.3742 H   0  0  0  0  0  0  0  0  0  0  0  0
    3.4061    0.7284   -3.4912 H   0  0  0  0  0  0  0  0  0  0  0  0
    1.8269   -0.2910   -3.6630 H   0  0  0  0  0  0  0  0  0  0  0  0
    2.3222   -2.4184   -1.1672 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.7122   -1.4701   -1.4491 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.7806   -2.1350    1.2247 H   0  0  0  0  0  0  0  0  0  0  0  0
    1.0672   -0.2707    1.1218 H   0  0  0  0  0  0  0  0  0  0  0  0
    3.6949   -1.2726    3.7391 H   0  0  0  0  0  0  0  0  0  0  0  0
    2.3186   -2.2231    3.6343 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  2  3  1  0  0  0  0
  3  4  1  0  0  0  0
  4  5  1  0  0  0  0
  5  6  1  0  0  0  0
  1  7  1  0  0  0  0
  1  8  1  0  0  0  0
  1  9  1  0  0  0  0
  3 10  1  0  0  0  0
  3 11  1  0  0  0  0
  4 12  1  0  0  0  0
  4 13  1  0  0  0  0
  6 14  1  0  0  0  0
  6 15  1  0  0  0  0
M  END
"
`;
